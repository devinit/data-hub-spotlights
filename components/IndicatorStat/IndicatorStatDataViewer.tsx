import React, { FunctionComponent, ReactText } from 'react';

interface IndicatorStatDataProps {
  note?: ReactText;
  value?: ReactText;
}

const IndicatorStatDataViewer: FunctionComponent<IndicatorStatDataProps> = ({ value, note }) => {
  return (
    <p className="spotlight__stat-data">
      {value}
      {note ? <span className="spotlight__stat-data__note">{note}</span> : null}
    </p>
  );
};

export { IndicatorStatDataViewer };
