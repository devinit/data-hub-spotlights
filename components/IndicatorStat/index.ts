export * from './IndicatorStat';
export * from './IndicatorStatDataHandler';
