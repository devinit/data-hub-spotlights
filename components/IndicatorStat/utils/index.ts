export * from './data';
export * from './chart';
