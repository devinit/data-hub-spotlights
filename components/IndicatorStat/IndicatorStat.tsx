import React, { FunctionComponent } from 'react';
import { ContentMeta } from '../../utils';

interface IndicatorStatProps {
  heading?: string;
  meta?: ContentMeta;
}
// TODO: add proper tooltip for description & source
const IndicatorStat: FunctionComponent<IndicatorStatProps> = ({ meta = {}, heading, children }) => {
  return (
    <div className="spotlight__stat">
      <h3 className="spotlight__stat-heading">
        {heading}
        {meta.description || meta.source ? (
          <span className="spotlight__stat-icon">
            <i role="presentation" aria-hidden="true" className="ico ico--12 ico-info-slate"></i>
            <style jsx>{`
              .spotlight__stat-icon {
                display: none;
              }
            `}</style>
          </span>
        ) : null}
      </h3>
      {children}
    </div>
  );
};

IndicatorStat.defaultProps = { meta: {} };

export { IndicatorStat };
