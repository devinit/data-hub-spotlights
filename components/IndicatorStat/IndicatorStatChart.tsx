import React, { FunctionComponent } from 'react';
import { IndicatorChart } from '../../utils';
import { EChartsBaseChart } from '../EChartsBaseChart';

type IndicatorStatChartProps = IndicatorChart;

const IndicatorStatChart: FunctionComponent<IndicatorStatChartProps> = ({ options }) => {
  return (
    <div className="spotlight__interactive">
      <EChartsBaseChart options={options} />
    </div>
  );
};

export { IndicatorStatChart };
