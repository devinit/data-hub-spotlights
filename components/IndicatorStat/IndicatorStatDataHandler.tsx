import React, { FunctionComponent } from 'react';
import { LocationIndicatorData } from '../../utils';
import { getIndicatorValue, ValueOptions, getIndicatorsValue } from '../KeyFactIndicator/utils';
import { IndicatorStatDataViewer } from './IndicatorStatDataViewer';

interface DataHandlerProps {
  data?: LocationIndicatorData[];
  dataLoading?: boolean;
  valueOptions?: ValueOptions;
}

const IndicatorStatDataHandler: FunctionComponent<DataHandlerProps> = ({ data, dataLoading, ...props }) => {
  if (!dataLoading && data) {
    if (data.length === 1) {
      return <IndicatorStatDataViewer value={getIndicatorValue(data[0].data, props.valueOptions)} />;
    }

    return <IndicatorStatDataViewer value={getIndicatorsValue(data, props.valueOptions)} />;
  }

  return <div>Loading ...</div>;
};

IndicatorStatDataHandler.defaultProps = { valueOptions: { dataFormat: 'plain' } };

export { IndicatorStatDataHandler };
