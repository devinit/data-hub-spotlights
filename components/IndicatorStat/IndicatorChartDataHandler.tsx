import React, { FunctionComponent, useEffect, useState } from 'react';
import { IndicatorChart } from '../../utils';
import { Alert } from '../Alert';
import { DataLoaderProps, useDDWData } from '../DDWDataLoader';
import { IndicatorStatChart } from './IndicatorStatChart';
import { generateChartOptions } from './utils';

interface DataHandlerProps extends IndicatorChart {
  dataOptions: DataLoaderProps;
}

const IndicatorChartDataHandler: FunctionComponent<DataHandlerProps> = ({ dataOptions, ...configs }) => {
  const [retryCount, setRetryCount] = useState(0);
  const { data, dataLoading, refetch, error } = useDDWData(dataOptions);
  useEffect(() => {
    if (error) {
      const { message } = error;
      if (message.includes('relation') && message.includes('does not exist') && retryCount === 0 && refetch) {
        refetch();
        setRetryCount(retryCount + 1);
      }
    } else {
      setRetryCount(0);
    }
  }, [error]);

  if (dataLoading) {
    return <div>Loading ...</div>;
  }

  if (error) {
    // TODO: add manual reload button
    return <Alert variant="error">Something went wrong while loading this widget</Alert>;
  }

  if (data) {
    if (data.length === 1) {
      return <IndicatorStatChart options={generateChartOptions(configs, data)} />;
    }
    console.log('TODO: handle charts built from multiple indicators');

    // TODO: handle charts built from multiple indicators
  }

  return <div>No Data</div>;
};

export { IndicatorChartDataHandler };
