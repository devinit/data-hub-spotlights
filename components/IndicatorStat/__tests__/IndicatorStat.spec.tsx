/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { IndicatorStat } from '../IndicatorStat';

describe('IndicatorStat', () => {
  test('renders default state correctly', () => {
    const renderer = TestRenderer.create(<IndicatorStat />).toJSON();

    expect(renderer).toMatchSnapshot();
  });
});
