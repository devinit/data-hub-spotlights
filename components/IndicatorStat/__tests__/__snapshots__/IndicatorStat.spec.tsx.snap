// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndicatorStat renders default state correctly 1`] = `
<div
  className="l-2up-3up__col"
>
  <div
    className="spotlight__stat"
  >
    <h3
      className="spotlight__stat-heading"
    />
    <p
      className="spotlight__stat-data"
    />
  </div>
</div>
`;
